Support for annotated getters

h-thurow opened the pull request:
IntelliJ can reverse engineer a database schema to JPA annotated entity classes what is very convenient. The problem is that it only annotates getters not fields. So it is preferable that SansOrm supports property access too, not only field access.

Property access has another advantage. It offers more control about the process of reading or writing field values, e. g. one can leverage JavaBean's PropertyChangeSupport.

JPA defines the `@Access` annotation to explicitly set field or property access per class or attribute. Also there is a default access type detecting process described in case there is no explicit access type set. SansOrm does now follow the specification herein.

I did not publish directly to origin/master in case you still want to try to clean up the last commit.

h-thurow commented:
What do you think about first deploying only to the staging repository? So we could ensure from there that there are no bugs. It is somehow embarrassing to release a buggy version to Maven Central.

h-thurow commented:
> When you’re satisfied with stability, you can request a release.

Your review is appreciated

brettwooldridge commented:
We can certainly publish snapshots to the snapshot repo.  Regarding testing, I can only verify that our usage is not broken, but that will not include method annotations.

h-thurow commented:
Our production test suite did also pass, but so far we applied only field annotation. We would start with property access and perhaps migrate existing field access to property access as soon as we could load the snapshot. So I would recommend to deploy a snapshot first. When after some time everything still looks good, we can release with confidence.

h-thurow commented:
Is it fine to merge and deploy a snapshot?

h-thurow commented:
There is still an issue with `@JoinColumn` annotations, I want to fix before deploying a snapshot.

h-thurow commented:
Could it be that self join is already broken in V 3.2? 

I have this entity with a self join field:

```
@Table(name = "JOINTEST")
public static class FieldAccessedOneToOneSelfJoin {
    @Id @GeneratedValue
    private int id;
    @JoinColumn(name = "parentId", referencedColumnName = "id")
    private FieldAccessedOneToOneSelfJoin parentId;
    private String type;
}
```

I have created this table:

```
    SqlClosureElf.executeUpdate(
        " CREATE TABLE JOINTEST (
           "id INTEGER NOT NULL IDENTITY PRIMARY KEY"
           + ", parentId INTEGER"
           + ", type VARCHAR(128)" +
           ", CONSTRAINT cnst1 FOREIGN KEY(parentId) REFERENCES (id)"
           + ")");
```
When I try to insert a parent with insertObject(...)

```
    FieldAccessedOneToOneSelfJoin parent = new FieldAccessedOneToOneSelfJoin();
        parent.type = "parent";
    SqlClosureElf.insertObject(parent);
```
I get
<pre>
org.h2.jdbc.JdbcSQLException: Syntax error in SQL statement "INSERT INTO JOINTEST) VALUES )"
</pre>
Maybe `@JoinColumn` is only supported with reads? So I inserted a parent and a child with executeUpdate(...)

```
    SqlClosureElf.executeUpdate("insert into JOINTEST (parentId, type) values (null, 'parent')");
    SqlClosureElf.executeUpdate("insert into JOINTEST (parentId, type) values (1, 'child')");
```
 
When I try to retrieve the child 

```
FieldAccessedOneToOneSelfJoin child = SqlClosureElf.objectFromClause
    (FieldAccessedOneToOneSelfJoin.class, "id=2");
```
I get
```
java.lang.IllegalArgumentException: Can not set 
    com.zaxxer.sansorm.internal.IntrospectedTest$FieldAccessedOneToOneSelfJoin.parentId to java.lang.Integer
```
I do not know how self joins were realized once upon a time.

When `@Access(PROPERTY)` is set I now implemented it complying with the spec: parentId is expected to be an Instance of the given field type to store parentId.
```
PropertyAccessedOneToOneSelfJoin parent = new PropertyAccessedOneToOneSelfJoin();
parent.type = "parent";
PropertyAccessedOneToOneSelfJoin child = new PropertyAccessedOneToOneSelfJoin();
child.type = "child";
child.parentId = parent;
SqlClosureElf.insertObject(child);
```
When child is read parentId again is retrieved as an instance of the given field type but with only the id field set, no other attributes. To retrieve the other attributes too one can call OrmElf.refresh(...) with the object as argument.

Would it break your usage when I implemented it this way with @Access(FIELD) too?

brettwooldridge commented on src/main/java/com/zaxxer/sansorm/internal/AttributeInfo.java:
```diff
@@ -10,59 +10,66 @@
 
-   private Class<?> getFieldType() {
-      final Class<?> type = field.getType();
+   protected abstract void extractFieldName(Field field);
```
It might seem nit-picky, but I'm all about using ``final`` on [parameters](https://github.com/brettwooldridge/SansOrm/blob/master/src/main/java/com/zaxxer/sansorm/internal/Introspected.java#L186) and [locals](https://github.com/brettwooldridge/SansOrm/blob/master/src/main/java/com/zaxxer/sansorm/internal/Introspected.java#L403).  Inadvertent modification can cause difficult to uncover issues and allows the IDE and compiler to catch such issues.

h-thurow replied:
I have now activated "Make generated local variables final" and "Make generated parameters final" in IntelliJ preferences.

brettwooldridge commented:
@h-thurow The self-join is probably the least documented feature.  Here is how our object (the only self-join in our whole system) is defined:
```java
@Entity(name = "ZHardware")
@Table(name = "device_hardware")
public class ZHardware {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "hardware_id")
    private int hardwareId;

    @Column(name = "device_id")
    private int deviceId;

    @ManyToOne(cascade = CascadeType.REFRESH)
    @JoinColumn(name = "parent_id", insertable = true, updatable = true)
    private ZHardware cardParent;
    ...
```
The ``@Entity`` and ``@ManyToOne`` annotations must be left-over from our Hibernate to SansOrm conversion. 😁 

The code that inserts the objects is:
```Java
SqlClosure.sqlExecute(connection -> {
   OrmElf.insertListNotBatched(connection, getCompletedStack());
   return null;
});
```
Where ``getCompletedStack()`` returns ``LinkedList<ZHardware>``.

h-thurow commented:
`@ManyToOne` would result in the field being ignored now. At least when access type is property.

brettwooldridge commented:
@h-thurow Is ``@ManyToOne`` even referenced in SansOrm?  The search of my (local) code turned up no usages.

brettwooldridge commented:
@h-thurow In reference to nothing in particular, I want to be careful not to turn SansOrm (which means "Without-ORM") into a full-blown ORM. 😁  I'm sure you will help keep SansOrm true to its roots, while adding sensible support for JPA annotations.

brettwooldridge commented:
@h-thurow Regarding *our* use of ``@ManyToOne``, I am happy to remove that annotation from our code (as it serves no purpose).  Looking at the code in your branch, it seems that would be sufficient to avoid any issues with the new code.  Is that a correct assessment?

brettwooldridge commented on src/main/java/com/zaxxer/sansorm/internal/AttributeInfo.java:
```diff
@@ -0,0 +1,358 @@
+            type.getCanonicalName() + " on "
+            + clazz.getCanonicalName());
+         toBeConsidered = false;
+      }
```
I'm not sure I agree with not throwing here.  It is a non-goal of SansOrm to allow "transparent" migration from objects annotated for a full ORM, for example Hibernate, to SansOrm.

I feel it would be better to clearly fail, such that users know what is definitely not supported by SansOrm.

While it may be possible to go from SanOrm -> "Full ORM", with respect to annotations, it is defintely not possible to go the other direction -- and is not a goal to do so as it is simply not possible unless SansOrm were in fact a full ORM.

brettwooldridge replied:
Note that that is *not* true of annotations that SansOrm does not support at all.  For example, ``@OneToMany`` may simply be ignored.  However, ``@JoinColumn`` *is* a *partially`` supported annotation, so we should throw when unsupported uses are discovered.

h-thurow replied:
GetterAnnotatedPitMainEntity shows an automatically with IntelliJ generated entity. We could not use it as-is with SansOrm because SansOrm will throw an Exception when processing
```
   @ManyToOne
   @JoinColumn(name = "PIT_IDENT", referencedColumnName = "PIR_PIT_IDENT", nullable = false)
   public GetterAnnotatedPitReferenceEntity getPitReferenceByPitIdent() {
      return pitReferenceByPitIdent;
   }
```

h-thurow commented:
I tested insertListNotBatched() in 3.2 with a single, really simple object:

   @Test
   public void selfJoinColumnH2insertListNotBatched() throws SQLException {

      JdbcDataSource ds = TestUtils.makeH2DataSource();
      SansOrm.initializeTxNone(ds);
      try (Connection con = ds.getConnection()){
         SqlClosureElf.executeUpdate(
            " CREATE TABLE JOINTEST ("
               + "id INTEGER NOT NULL IDENTITY PRIMARY KEY"
               + ", parentId INTEGER"
               + ", type VARCHAR(128)" +
               ", CONSTRAINT cnst1 FOREIGN KEY(parentId) REFERENCES (id)"
               + ")");

         FieldAccessedOneToOneSelfJoin parent = new FieldAccessedOneToOneSelfJoin();
         parent.type = "parent";
         ArrayList<FieldAccessedOneToOneSelfJoin> parents = new ArrayList<>();
         parents.add(parent);
         OrmElf.insertListNotBatched(con, parents); // throws Exception
      }
      finally {
         SqlClosureElf.executeUpdate("DROP TABLE JOINTEST");
      }
   }

Again this Exception is thrown:

    org.h2.jdbc.JdbcSQLException: Syntax error in SQL statement "INSERT INTO JOINTEST)[*] VALUES )"

So I think one can say it does not work anymore or at least it does not work as user would expect it to work.

h-thurow commented:
> "... `@ManyToOne`, I am happy to remove that annotation ... Looking at the code in your branch, it seems that would be sufficient to avoid any issues with the new code. Is that a correct assessment?"

Would it not be ambiguous? `@JoinColumn` alone does not describe the kind of association. There is no hint in the spec that it defaults to a specific relationship.

brettwooldridge commented:
Ultimately, SansOrm was created to move *away* from ORMs.  It was intended to be a one-way trip.

We considered not using JPA annotations at all, and simply defining our own, much as Jackson defined their own JSON annotations.

In the end, we decided to just re-purpise existing annotations for expediency.  Adhering to the “spec” is not especially important, and if it leads to confusion, we should just move away from JPA annotations.

Frankly, annotations like ``@ManyToOne`` have no meaning for SansOrm, because SansOrm does not do “relationships”.  We didn’t even need to use ``@JoinColumn``,, we could have just used ``@Column``, as the self-association could be inferred by the type of the property.

h-thurow commented:
Too bad, this intention is to restrictive to us.

brettwooldridge commented:
@h-thurow Thank you for your contributions.  I'm sorry if that came as a surprise, I thought we were pretty clear in the first paragraph of the project page:

> [SansOrm] was created to rid my company's product of Hibernate. After about 10 years of using ORMs in various projects, I came to the same conclusion as others: [ORM is an Anti-Pattern.](https://github.com/brettwooldridge/SansOrm/wiki/ORM-is-an-anti-pattern)

h-thurow commented:
That it is not an OR mapper is the attracting thing. But I do not feel well in a one-way street. Our central classes shall not be tied  to a specific frame work or restricted by it. I already knew "ORM is an Anti-Pattern" long time before I met SansOrm.
